networkd: add support to configure CAN devices

marckleinebudde opened the pull request:
This patch series adds systemd-networkd support for kind "can". Fixes: https://github.com/systemd/systemd/issues/4042

poettering commented on src/network/networkd-link.c:
```diff
@@ -1872,6 +1873,86 @@ static int link_up_can(Link *link) {
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not append IFLA_INFO_DATA attribute: %m");
+
+        if (link->network->can_bitrate || link->network->can_sample_point) {
```
we generally prefer if numeric variables are check for with explicit comparisons instead of relying on C's downgrade-to-bool feature for them. I.e. unless something is a bool or is bool-like (or a pointer), please write explicitly "if (x != 0)" rather than just "if (x)"

marckleinebudde replied:
OK

poettering commented on src/network/networkd-link.c:
```diff
@@ -1872,6 +1873,86 @@ static int link_up_can(Link *link) {
+                               format_timespan(time_string, FORMAT_TIMESPAN_MAX,
+                                               link->network->can_restart_us, USEC_PER_SEC));
+
+                r = sd_netlink_message_append_u32(m, IFLA_CAN_RESTART_MS, link->network->can_restart_us / USEC_PER_MSEC);
```
i don't know CAN, but wouldn't it make more sense to pass DIV_ROUND_UP(link->network->can_restart_us, USEC_PER_MSEC) here? i.e. round up rather down when converting from µs to ms?

marckleinebudde replied:
Will use DIV_ROUND_UP.

poettering commented on src/network/networkd-link.c:
```diff
@@ -2628,7 +2729,7 @@ static int link_configure(Link *link) {
         assert(link->state == LINK_STATE_PENDING);
 
-        if (streq_ptr(link->kind, "vcan"))
+        if (streq_ptr(link->kind, "vcan") || streq_ptr(link->kind, "can"))
```
`STRPTR_IN_SET(link->kind, "can", "vcan")` please

marckleinebudde replied:
nice helper!

poettering commented on src/network/networkd-link.c:
```diff
@@ -1872,6 +1873,86 @@ static int link_up_can(Link *link) {
+                        return log_link_error_errno(link, r, "Could not append IFLA_CAN_BITTIMING attribute: %m");
+        }
+
+        if (link->network->can_restart_us) {
```
as above, please use explicit numeric comparisons for this...

Hmm, here's one question though: we try to be careful with time value of 0 these days. Some APIs are sloppy and say that a time value of zero equals "off", even though a time value of zero could conceptually also mean "right-away". Note that our usec parsers usually also accept parsing the string "infinity" into the special value USEC_INFINITY (which is (usec_t) -1)). Quite often allowing a zero value while saying "infinity" means "off" is the better approach. Is this a case like this?

marckleinebudde replied:
Then `restart_ms` is one these sloppy APIs. `restart_ms == 0` means never automatically restart, `USEC_INFINITY` would mean the same.

poettering commented on src/network/networkd-network-gperf.gperf:
```diff
@@ -178,6 +178,9 @@ IPv6Prefix.OnLink,                      config_parse_prefix_flags,
 IPv6Prefix.ValidLifetimeSec,            config_parse_prefix_lifetime,                   0,                             0
 IPv6Prefix.PreferredLifetimeSec,        config_parse_prefix_lifetime,                   0,                             0
+CAN.BitRate,                            config_parse_si_size,                           0,                             offsetof(Network, can_bitrate)
+CAN.SamplePoint,                        config_parse_uint32,                            0,                             offsetof(Network, can_sample_point)
```
hmm, config_parse_uint32 is a bit sloppy here, no? I mean, i presume the SamplePoint= field only takes a value in the range 0…1000 or so? please add a more precise parser here...

also, iiuc this is a permille value, right? if so, please consider adding parse_permille() like we already have parse_percent(), and then match this up with config_parse_permille() too...

parse_permille() should probably accept values such as "55‰". But given that that involves unicode, an alternative syntax with a single separating dot followed by a single digit might make sense to support as well, i.e. "5.5%"?

i.e. something along the lines of:

```c
int parse_permille_unbounded(const char *p) {
        const char *pc, *pm, *n;
        int r, v;

        pm = endswith(p, "‰");
        if (pm) {
                n = strndupa(p, pm - p);
                r = safe_atoi(n, &v);
                if (r < 0)
                        return r;
        } else {
                pc = endswith(p, "%");
                if (!pc)
                        return -EINVAL;

                dot = memchr(p, pc - p, '.');
                if (dot) {
                        if (dot + 2 != pc)
                                  return -EINVAL;
                        if (dot[1] < '0' || dot[1] > '9')
                                  return -EINVAL;
                        q = dot[1] - '0';
                        n = strndupa(p, dot - p);
                } else {
                        q = 0;
                        n = strndupa(p, pc - p);
                }
                r = safe_atoi(n, &v);
                if (r < 0)
                        return r;
                if (v * 10 + q < v) /* overflow */
                        return -ERANGE;

                v = v * 10 + q;
        }
        if (v < 0)
                return -ERANGE;

        return v;
}

int parse_permille(const char *p) {
        int v;

        v = parse_permilled_unbounded(p);
        if (v > 1000)
                return -ERANGE;

        return v;
}
```

(this is entirely untested, but you get the idea)

marckleinebudde replied:
Works now, I've also added some test.

poettering commented on src/network/networkd-network.h:
```diff
@@ -191,6 +191,11 @@ struct Network {
 
+        /* CAN support */
+        uint32_t can_bitrate;
+        uint32_t can_sample_point;
```
hmm, config_parse_permille() currently expects to write to an "int" location, you define this here as "uint32_t"... Of course on all systems we care about the difference doesn't matter, given that "int" is 32bit, and we know that the parsed value is between 0…1000. However, I think we should be more careful here, and not access "uint32" data fields through "int" pointers.

I'd probably settle for using the "unsigned" type here. i.e. change config_parse_permille() to cast the return value of parse_permille() to an unsigned after handling the errors it might return, and then assigning it to an unsigned pointer. And hence change the field in the Network structure to an unsigned, too. Why "unsigned"? Well, it's relatively generic, and covers the 0…1000 range pretty closely, which is all we need here...

marckleinebudde replied:
`parse_permille()` does return an `int`, with `-errno` in case of error and a value from `0...1000` otherwise. `config_parse_permille()` will ignore the any errors (as most `config_parse_*()` do), so I'll make `struct Network::can_cample_point` an int. In `networkd-link.c` `link_set_can()` I'll trust that `network->can_sample_point` is `0...1000` and convert it without further checks into an `uint32_t`. Is this OK?

poettering replied:
Well, all other variables we store percent values in so far used "unsigned". Hence I'd really pick the same type for permille too, even if it means casting from int to unsigned after handling the parse error.

I mean, this doesn't matter much, it's mostly a question of internal consistency, and least surprises.

I mean, if you store a percent/permille value in an `unsigned`, then any value you could assign it to makes sense. if you store it in a simple `int`, then things are a bit weird because negative perent/permille are really weird in almost all cases...

But dunno, this is not a biggie

marckleinebudde replied:
Using `unsigned` now.

poettering commented on src/network/networkd-link.c:
```diff
@@ -1872,6 +1873,101 @@ static int link_up_can(Link *link) {
+
+                r = sd_netlink_message_append_u32(m, IFLA_CAN_RESTART_MS, restart_ms);
+
+                if (r < 0)
```
pleae drop the empty line, we try to keep function calls and their error handling together

marckleinebudde replied:
ok

poettering commented on man/systemd.network.xml:
```diff
@@ -1753,6 +1753,35 @@
+          <listitem>
+            <para>Automatic restart delay time. If set to a non-zero value, a restart of the CAN controller will be
+            triggered automatically in case of a bus-off condition after the specified delay time. By default it's
+            off.</para>
```
should say that "RestartSec=infinity" is what turns it off

marckleinebudde replied:
ok

poettering commented on man/systemd.network.xml:
```diff
@@ -1753,6 +1753,35 @@
         </varlistentry>
       </variablelist>
```
this commit should carry "Fixes: #4042" in the commit text so that github can automatically close the issue when this is merged.

marckleinebudde replied:
ok

poettering commented on man/systemd.network.xml:
```diff
@@ -1753,6 +1753,35 @@
+        <varlistentry>
+          <term><varname>BitRate=</varname></term>
+          <listitem>
+            <para>The bitrate of CAN device in bit per second.</para>
```
might be worth saying that the usual  K, M, G, … suffixes are accepted, to the base of 1000

marckleinebudde replied:
ok

AlexanderStein commented on src/network/networkd-network-gperf.gperf:
```diff
@@ -178,6 +178,9 @@ IPv6Prefix.OnLink,                      config_parse_prefix_flags,
 IPv6Prefix.PreferredLifetimeSec,        config_parse_prefix_lifetime,                   0,                             0
+CAN.BitRate,                            config_parse_si_size,                           0,                             offsetof(Network, can_bitrate)
+CAN.SamplePoint,                        config_parse_permille,                          0,                             offsetof(Network, can_sample_point)
+CAN.RestartSec,                         config_parse_sec,                               0,                             offsetof(Network, can_restart_us)
```
Shouldn't this be called CAN.RestartSec? Reading this I was wondering how can I set times with sub second precision until I figured the usage.

marckleinebudde replied:
It _is_ called `CAN.RestartSec`. I grep'ed over the systemd source an all config options using the `config_parse_sec` parser end with `Sec`. However there is a `config_parse_nsec` parser and these options use `NSec`. The restart delay in the CAN world is usually some 100 micro seconds.

marckleinebudde replied:
The kernel interface is using microseconds, but IMHO that's not important here. I'll mention in the man page how set a timeout of 100ms.

marckleinebudde commented on src/network/networkd-link.c:
```diff
@@ -1872,6 +1873,101 @@ static int link_up_can(Link *link) {
+        if (r < 0)
+                return log_link_error_errno(link, r, "Could not append IFLA_INFO_DATA attribute: %m");
+
+        if (link->network->can_bitrate > 0 || link->network->can_sample_point > 0) {
```
The kernel will refuse to bring up the interface:
```
Jun 08 10:57:30 hardanger kernel: gs_usb 2-2:1.0 can0: bit-timing not yet defined
```
While `systemd-networkd` logs:
```
can0: link_set_can
can0: Setting bitrate = 0 bit/s
can0: Setting sample point = 72.5%
can0: Bringing CAN link up
can0: link_set_can done
can0: Set link
can0: Could not join netdev: Invalid argument
can0: Failed
```

poettering commented:
excellent! thanks!

keszybz commented on src/basic/parse-util.c:
```diff
@@ -632,6 +632,58 @@ int parse_percent(const char *p) {
+                r = safe_atoi(n, &v);
+                if (r < 0)
+                        return r;
+                if (v * 10 + q < v) /* overflow */
```
That's signed overflow, undefined behaviour.

But why not just use `strtod`?

marckleinebudde replied:
The `int parse_permille_unbounded()` was written and suggested by @poettering .

keszybz replied:
Still a bug ;)
See e.g. https://wiki.sei.cmu.edu/confluence/display/c/INT32-C.+Ensure+that+operations+on+signed+integers+do+not+result+in+overflow for a nice explanation.

poettering replied:
strtod() is never Ok to use, because it is locale dependent, which we really don't want here... (dot vs. comma...)

keszybz commented on man/systemd.network.xml:
```diff
@@ -1753,6 +1753,40 @@
+  <refsect1>
+    <title>[CAN] Section Options</title>
+      <para>The <literal>[CAN]</literal> section manages the Controller Area Network (CAN bus) accepts
+      the following keys.</para>
```
"... manages ... accepts" — there are two verbs in this sentence.

marckleinebudde replied:
...and accepts...

keszybz commented on man/systemd.network.xml:
```diff
@@ -1753,6 +1753,40 @@
+        <varlistentry>
+          <term><varname>BitRate=</varname></term>
+          <listitem>
+            <para>The bitrate of CAN device in bit per second. The usual si prefixes (K, M) with the base of 1000 can
```
bit**s** per second

si → SI

marckleinebudde replied:
ok

keszybz commented on man/systemd.network.xml:
```diff
@@ -1753,6 +1753,40 @@
+          <term><varname>RestartSec=</varname></term>
+          <listitem>
+            <para>Automatic restart delay time. If set to a non-zero value, a restart of the CAN controller will be
+            triggered automatically in case of a bus-off condition after the specified delay time. Sub second delays
```
sub second → subsecond

marckleinebudde replied:
ok

keszybz commented on man/systemd.network.xml:
```diff
@@ -1753,6 +1753,40 @@
+            triggered automatically in case of a bus-off condition after the specified delay time. Sub second delays
+            can be specified using decimals (e.g. <literal>0.1s</literal>) or a <literal>ms</literal> or
+            <literal>us</literal> postfix. Using <literal>infinity</literal> or <literal>0</literal> will turn the
+            automatic restart off. When omitting the parameter <varname>RestartSec=</varname> automatic restart is
```
"When omitting …" is awkwared. Maybe just "By default automatic restart is disabled." ?

marckleinebudde replied:
Yes, reads much more fluent that way.

marckleinebudde commented:
I've pushed the documentation changes. Let's wait for @poettering to comment on `int parse_permille_unbounded()`.

poettering commented:
maybe use this:  

```c
if ((unsigned) v * 10U + (unsigned) q > (unsigned) INT_MAX)
        return -ERANGE;
```

This should be safe, as "UINT_MAX > INT_MAX*10 + 10" is generally true.

`

marckleinebudde commented:
This work, too:
```c
                if (v > ((INT_MAX / 10) + q))
                        return -ERANGE;

                v = v * 10 + q;
```

```
parse_permille ----> 214748364.7%: 2147483647 7fffffff
parse_permille ----> 214748364.8%: -34 ffffffde
```

marckleinebudde commented:
I've changed the code to:

```c
                if (v > ((INT_MAX / 10) + q))
                        return -ERANGE;

                v = v * 10 + q;
```
and it works like this:
```
parse_permille ----> 214748364.7%: 2147483647 7fffffff
parse_permille ----> 214748364.8%: -34 ffffffde
```
I've added test a overflow test to the testsuite.

marckleinebudde commented:
BTW:
```c
if ((unsigned) v * 10U + (unsigned) q > (unsigned) INT_MAX)
        return -ERANGE;
```
fails with:
```
Assertion 'parse_permille_unbounded("429496729.6%") == -ERANGE' failed at ../src/test/test-parse-util.c:694, function test_parse_permille_unbounded(). Aborting.
```

marckleinebudde commented:
BTW:
```c
if (v > ((INT_MAX / 10) + q))
        return -ERANGE;
```
this is also broken, but the
```c
if (v < 0)
        return -ERANGE;
```
caught it. This looks promising:
```c
                if (v > ((INT_MAX - q) / 10))
                        return -ERANGE;

                v = v * 10 + q;
```

poettering commented:
Excellent! Thanks for doing the overflow work, much appreciated. Merging.

keszybz commented:
Yep, looks nice.

trabant-rgb commented:
Just learned that I could manage my socket can interface with systemd-networkd as I do with my ip interfaces. This may or may not be the correct place to ask... I didn't find a parameter to set the can interface to "listen-only" mode. My current way is to bring the can0 up within a systemd .service as 

[Service]
ExecStartPre=/sbin/ip link set can0 up type can bitrate 10000 listen-only on
ExecStart=/home/pi/bin/can-vz.sh

... however it would be much cleaner to do this in a .network file. Any chance to get that option working?

Thanks,
Alex

boucman commented:
please open a new feature-request ? ideally even do a pull-request if you feel like coding it...
